### ISSUE TITLE
Options generation annotations on entities

### ISSUE THREAD

**Issue by Tcharl** [jhipster/generator-jhipster#26007]
##### **Overview of the feature request**

The goal would be to add some jdl annotations and cli options on entities.

The goal would be to split annotations into two categories: layer ones vs C-R-U-D ones.

-- Frontend --
* `@Form` would generate frontend forms
* `@FrontConsumer` would generate client-side consumer no matter the consumption strategy

-- Controllers --
* `@Controller(types=Rest)` would expose api
* `@Controller(types=Sse)` would expose streaming (listeners and senders)
* `@Consumer(types=Stream[<entitytopic>])` would Consume message via streams (data coming from other microservice)

-- Services --

* `@Consumer(types=Rest[<toApplication>/<toEntity>])` would call the target service api instead of repository in the service implementation.
* `@Consumerr(types=Stream[<topics>])` would hook service.save() and send messages in the topics instead of using repositories
* `@Consumer(types=SQL)`would use usual spring data providers

-- C-R-U-D --

Same logic to be applied to methods (create:update/patch/get/delete), decorelated to layers.

--- summing  up
As such, 'reasonable default' as per [Inclusive property rfc](https://github.com/jhipster/generator-jhipster/blob/main/rfcs/5-jhipster-rfc-inclusive-property-in-generator.md) would be:

```
@Form
@FrontConsumer
@Controller(types=Rest)
@Consumer(types=Repository)
entity <name>{ ... }
```

and for crud:
```
// Won't generate the 'delete button', neither the methods
@CREATE
@GET
@LIST
@UPDATE
entity <name>{ ... }
```

And to let user be more precise by combining both (overriding class-level choices):
```
@Form(verbs=GET,UPDATE) // this choice won't generate the 'list', 'creation' and 'deletion' form
@FrontService(types=Sse) // will consume the Sse frontconsumer
@FrontConsumer // Will generate both consumer
@Controller(types=Sse,REST,Stream[topic=<name>])
@Service(mode=DTO)
@Consumer(types=SQL)
@Repository(verbs=CREATE,GET,LIST,UPDATE) // Superseeds the annotations below
@CREATE
@GET
@UPDATE
entity <name>{ ... }

```

##### **Motivation for or Use Case**

The goal would be to provide much more control on entities generation

The ultimate goal being to provide a versatile solution for complex applications with few effort from the end user (plumbery jhipster-managed), i.e.

```
application clientCartPortal {
  entities: PurchaseRequest
  type: gateway
}
application supplySink {
  entities: Supply,Order
  type: microservice
}
application supplyOperatorPortal {
  entities: Shipping,OrderStatus
  type: gateway
}

@Form
@FrontConsumer
@Controller(types=Rest, Stream[orders])
@CREATE
@UPDATE
@GET
entity PurchaseRequest extends Supply {} // can just replicate the fields while extensions not supported

@Controller(types=Stream[shippedsuplies]) // user will just have to filter element with status 'shipped' to the stream
@Consumer(types=SQL, Stream[orders])
entity Supply {
  quantity int
  itemId UUID // or relations to stock.item
  unitprice float
}

entity OrderStatus {
  status: enum [Waiting,Processing,Shipped]
}

oneToMany {
Supply{status} to OrderStatus{supplies}
}

@Form
@FrontConsumer
@Controller(types=Sse)
@Consumer(types=Streams[shippedsuplies])
@LIST
entity Shipping extends Supply {} // shipping portal will receive updates ;-)
```

I want to contribute to this feature, however, seems huge so need some workforce help + bounties please (to motivate other contributors)!
@qmonmert @mshima @mraible @MathieuAA wdyt ?

- [ x ] Checking this box is mandatory (this is just to show you read everything)

**Comment by MathieuAA** [comment]
Hello everyone, while I don't doubt it's doable right now, there may first have to be a design phase so as to make the necessary adjustments to the parsing system, then another change in the generator.
The JDL parsing system identifies what's relevant, map and store it in some way, then creates objects (different from JHipster object shapes) based on what has been previously read.
Why it is relevant: this intermediary step is necessary so as not to be tightly bound to the generator. And yes, this requires a last mapping to get JHipster shapes, this is by design and should not be changed IMO.
What happens without it: one makes changes to the generator system, then makes a similar change to the parsing system (here be binding). Now say one wants to change the generator in some fashion, and forgets to change the parsing system accordingly. The next time someone wants to change both, there will be a difference between the two and a confusion: what version is right? (That's why tests are important.)

To sum up, to bind every annotation to a specific generation, my opinion is that one has to: make a change to the parsing system to acknowledge said change, wire it to the generator with an **explicit** mapping (we leave the parsing system to enter the generator system, this gate matters), and then read the output of this mapping to generate what's wanted.

**Comment by mshima** [comment]
I think `entity` does not apply to this.
Maybe it should be `model`.
`entity` would be a shortcut for `model` with many optionals.

We currently have support for `@EntityDomainLayer(false)`, `@EntityRestLayer(false)`, `@EntityPersistenceLayer(false)`, `@EntitySearchLayer(false)`

**Comment by Tcharl** [comment]
Hi @mshima , wasn't aware about this concept, totally make sense!

First question: do you know if the current supported annotations are operationally used or beta? According to RFC, reasonable default should be used if nothing is specified, otherwise any annotation superseeds default.

Thus, specifying 
```
entity foo {}
```
would be equals to 
```
@EntityForm
@EntityFrontRestClient
@EntityRestLayer // would prefer producer for consistency
@EntityService(interface=true) // compared to serviceClass
@EntityDTO
@EntityPersistenceLayer
@EntityDomainLayer
```

Meaning that the `(false)` option shouldn't exist, and IMHO allowlist provides much more capabilities/flexibility than denylist


For the model vs entity, I definitely love the idea for different reason.

From what I understand, there's a decoupling between the model (description of the class, logical layer) and the entity which is the real instance of the class.

Let's illustrate pseudocoding:

```
application operatorPortal {
  name: operator
  type sql
  entities: Supply, ShippingStatus
}
application customerPortal {
  name: customer
  type elastic
  entities: Order, PaymentStatus
}

model Goods {
  itemId: UUID
  quantity: number
  address: string
}

@EntityForm
@EntityFrontRestClient
@EntityRestLayer // would prefer producer for consistency
@EntityService(interface=true) // compared to serviceClass
@EntityDTO
@RestClient(targetApplication=operatorPortal, targetEntity=Supply)
entity Order model Goods { // Only model (Goods) attributes will be used for transfert
  status: PaymentStatus
}



@EntityRestLayer // would prefer producer for consistency
@EntityService(interface=true) // compared to serviceClass
@EntityDTO
@EntityPersistenceLayer
@EntityDomainLayer
entity Supply model Goods {
  status: ShippingStatus
}

```

Any remark?

**Comment by Tcharl** [comment]
Please keep opened
